File: debug-skill/scripts/debug_server.py
#!/usr/bin/env python3
"""
Debug Log Server

Receives logs via HTTP POST and writes them to {project}/.claude/debug-{sessionId}.log
Supports multiple concurrent sessions via sessionId parameter.

Usage:
    python3 debug_server.py /path/to/project
    python3 debug_server.py  # Uses current directory
"""
from http.server import HTTPServer, BaseHTTPRequestHandler
import json
import os
import sys
from datetime import datetime

LOG_DIR = sys.argv[1] if len(sys.argv) > 1 else "."
CLAUDE_DIR = os.path.join(LOG_DIR, ".claude")
PORT = int(os.environ.get("DEBUG_PORT", "8787"))

os.makedirs(CLAUDE_DIR, exist_ok=True)


def get_log_file(session_id):
    return os.path.join(CLAUDE_DIR, f"debug-{session_id}.log")


class DebugHandler(BaseHTTPRequestHandler):
    def log_message(self, *args):
        pass  # Suppress default logging

    def _send_cors_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def do_OPTIONS(self):
        self.send_response(204)
        self._send_cors_headers()
        self.end_headers()

    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self._send_cors_headers()
        self.end_headers()
        self.wfile.write(json.dumps({"status": "ok", "log_dir": CLAUDE_DIR}).encode())

    def do_POST(self):
        try:
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length)
            data = json.loads(body) if body else {}

            session_id = data.pop("sessionId", "default")
            log_file = get_log_file(session_id)

            entry = {
                "ts": datetime.now().isoformat(),
                **data
            }

            with open(log_file, "a") as f:
                f.write(json.dumps(entry) + "\n")

            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self._send_cors_headers()
            self.end_headers()
            self.wfile.write(json.dumps({"ok": True, "log_file": log_file}).encode())

            msg = entry.get("msg", json.dumps(entry)[:80])
            print(f"[{session_id}] {msg}")

        except Exception as e:
            self.send_response(400)
            self.send_header("Content-Type", "application/json")
            self._send_cors_headers()
            self.end_headers()
            self.wfile.write(json.dumps({"error": str(e)}).encode())


def main():
    print(f"Debug Log Server")
    print(f"  Endpoint: http://localhost:{PORT}/log")
    print(f"  Log dir:  {CLAUDE_DIR}")
    print(f"  Press Ctrl+C to stop\n")

    server = HTTPServer(("", PORT), DebugHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nServer stopped.")


if __name__ == "__main__":
    main()

File: debug-skill/scripts/debug_server.js
#!/usr/bin/env node
/**
 * Debug Log Server
 *
 * Receives logs via HTTP POST and writes them to {project}/.debug/debug-{sessionId}.log
 * Supports multiple concurrent sessions via sessionId parameter.
 *
 * Usage:
 *     node debug_server.js /path/to/project SESSION_ID
 *     node debug_server.js /path/to/project  # Error: session ID required
 *
 * Environment:
 *     DEBUG_LOG_DIR - Override log subdirectory (default: .debug)
 *     DEBUG_PORT    - Override port (default: 8787)
 */
const http = require('http');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const PROJECT_DIR = process.argv[2] || '.';
const SESSION_ID = process.argv[3];
const LOG_SUBDIR = process.env.DEBUG_LOG_DIR || '.debug';
const LOG_DIR = path.join(PROJECT_DIR, LOG_SUBDIR);
const PORT = parseInt(process.env.DEBUG_PORT || '8787', 10);

// Validate session ID is provided
if (!SESSION_ID) {
  console.error('Error: Session ID required');
  console.error('Usage: node debug_server.js /path/to/project SESSION_ID');
  console.error('');
  console.error('Generate session ID first:');
  console.error('  SESSION_ID="fix-description-$(uuidgen | cut -c1-6 | tr \'[:upper:]\' \'[:lower:]\')"');
  process.exit(1);
}

// Check if server already running
(async () => {
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 500);
    const res = await fetch(`http://localhost:${PORT}`, { signal: controller.signal });
    clearTimeout(timeout);
    if (res.ok) {
      console.log(`Server already running on port ${PORT}`);
      process.exit(0);
    }
  } catch {}

  // Cleanup stale port (crash recovery)
  try {
    const pid = execSync(`lsof -ti:${PORT}`, { encoding: 'utf-8' }).trim();
    if (pid) {
      console.log(`Cleaning up stale process on port ${PORT} (PID: ${pid})`);
      execSync(`kill -9 ${pid}`);
    }
  } catch {}

  startServer();
})();

function startServer() {
  fs.mkdirSync(LOG_DIR, { recursive: true });

  const getLogFile = (sessionId) => path.join(LOG_DIR, `debug-${sessionId}.log`);

  const server = http.createServer((req, res) => {
    const cors = {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type',
    };

    if (req.method === 'OPTIONS') {
      res.writeHead(204, cors);
      res.end();
      return;
    }

    if (req.method === 'GET') {
      res.writeHead(200, { 'Content-Type': 'application/json', ...cors });
      res.end(JSON.stringify({
        status: 'ok',
        session_id: SESSION_ID,
        log_file: path.join(LOG_DIR, `debug-${SESSION_ID}.log`),
      }));
      return;
    }

    if (req.method === 'POST' && req.url === '/log') {
      let body = '';
      req.on('data', chunk => (body += chunk));
      req.on('end', () => {
        try {
          const data = body ? JSON.parse(body) : {};
          const sessionId = data.sessionId || 'default';
          const logFile = getLogFile(sessionId);

          // Remove sessionId from stored entry (it's in filename)
          const { sessionId: _, ...rest } = data;
          const entry = { ts: new Date().toISOString(), ...rest };

          fs.appendFileSync(logFile, JSON.stringify(entry) + '\n');

          res.writeHead(200, { 'Content-Type': 'application/json', ...cors });
          res.end(JSON.stringify({ ok: true, log_file: logFile }));

          console.log(`[${sessionId}] ${entry.msg || JSON.stringify(entry).slice(0, 80)}`);
        } catch (e) {
          res.writeHead(400, { 'Content-Type': 'application/json', ...cors });
          res.end(JSON.stringify({ error: e.message }));
        }
      });
      return;
    }

    res.writeHead(404, cors);
    res.end('Not found');
  });

  server.listen(PORT, () => {
    console.log('Debug Log Server');
    console.log(`  Session:  ${SESSION_ID}`);
    console.log(`  Endpoint: http://localhost:${PORT}/log`);
    console.log(`  Log file: ${path.join(LOG_DIR, `debug-${SESSION_ID}.log`)}`);
    console.log('\nReady');
    console.log('\nPress Ctrl+C to stop');
  });
}
